### ISSUE TITLE
Is there any desire to fix the output structure of the codegen output?

### ISSUE THREAD

**Issue by sheldonj** [stephenh/joist-ts#967]
Loving this library, but really struggling with the structure that the generated code outputs. 

### Main Concerns that need addressing

- **There are tons of circular imports** -- this really is the main point.
- export * from 
- Lack of folder structure separating entities, enums, and other artifacts (metadata, entities). 
- No separation of generated files vs files meant to be extended by the consumer. 
-"any" typescript usage.

I've already started on creating a new codegen implementation for my internal use but was wondering if there was a joist opinion on where you would like this to land.  

I imagine there are a lot of things I don't know as I'm pretty new to this library. But I would be happy to own this work.

**Comment by stephenh** [comment]
Hi @sheldonj ! 

Yeah, coincidentally, we were just talking last week about re-organizing some of the output, mostly around simple moving of things i.e.:

* Put all factories into `src/entities/factories/*``
* Put all codegen files into `src/entities/codegen/*`

Granted, that's mostly a cosmetic change, just to reduce the number of files living directly in `src/entities`.

The `metadata.ts` and `entities.ts` we could also move into a `/codegen/` subdirectory.

Per the other points like:

> tons of circular imports

True, but I'm not sure there is a way to avoid that? Open to ideas, but due to:

a) JS classes need their base class defined before `class Author extends AuthorCodegen` is evaled (at import/require time, not "new Author()" time), and
b) domain models fundamentally have a lot of "author imports book, book imports author", i.e. the two-sided-ness of relations (which is one of joist's main features, modeling both of those & keeping both of those in sync)

I'm not sure there is really a way to avoid them, but again open to ideas or specific instances of circular imports that you think are avoidable.

(Fwiw [this post](https://medium.com/visual-development/how-to-fix-nasty-circular-dependency-issues-once-and-for-all-in-javascript-typescript-a04c987cf0de) is the approach Joist uses to avoid runtime issues of `class Author extends undefined` happening if somehow `Author` gets imported before `AuthorCodegen`--which will definitely happen, b/c we did not originally using this post's approach, and had to move to it to avoid the runtime issues it refers to, once a domain model gets even slightly large, non-deterministic import order & these errors are bound to happen imo / from our experience.)

> `export *`

Definitely open to more precise exports, but would need to evaluate that on a case-by-case basis.

> `any` usage

Would need to see more details/specific usages; I am personally comfortable using `any` within the guts of the Joist infra/implementation, but in theory the domain model / API / entities that Joist exposes to the application code itself should definitely not having any (?) `any` usages, at least as an idealistic goal, with maybe the occasional pragmatic exception.

**Comment by brudil** [comment]
Just to chime in on the codegen front, from our perspective it would be great to have all codegen'd files isolated in directories. We use Rush, which has a build cache feature but operates on caching directories (overwrite) rather than filename patterns. Currently we're unable to include Joist's process in the cached build process due to this.

**Comment by stephenh** [comment]
Hi @brudil ! Good to know re Rush. I'm curious, is it fine if the `/codegen/` directory has imports into entities? i.e. particular for CTI we could have:

* `src/entities/codegen/PublisherCodegen.ts`
* `src/entities/Publisher.ts` (`extends PublisherCodegen`)
* `src/entities/codegen/LargePublisherCodegen.ts` (`extends Publisher`)
* `src/entities/LargePublisher.ts` (`extends LargePublisherCodegen`)

So like if `tsc` tied to compiled `codegen/*` it may have dependencies back into `src/entities/*` and so the `codegen/*` directory would not be a standalone-compilable unit.

Will Rush's build caching care about that? Or will it not matter?

Either way, I'll start poking around at moving codegen files into `src/entities/codegen/*`. It's easy enough to do (I can't think of any reason why it wouldn't work...) but I'd also like to extend the codemod infra to auto-delete the old codegen files as users upgrade/install the new Joist version.

**Comment by brudil** [comment]
I haven't pushed forward on it too much as codegen isn't a particually common or slow process. The driving force/goal would be saving time in CI where we won't need to spin up a pg instance unless the migrations have changed.

My current thinking is it would work as follows:
- We have our migrations in a seperate package, @x/migrations. These are kysely migrations which are packaged and loaded with a custom MigrationProvider.
- Our service has a dependancy on these migrations
- Rush includes a package's deps in it's state hash

Rush is pretty agnostic in terms of what it caches. You purely define the script name and it's output directories; contents isn't read.

```json5
{
    "operationName": "codegen",
    "outputFolderNames": ["src/entities/codegen"],
	"dependsOnAdditionalFiles": ["joist-codegen.json"]
  },
  {
    "operationName": "test",
    "outputFolderNames": ["coverage"]
  }
 // ... etc
```

So I think with the pipeline looking like codegen, _then_ build. This should work pretty well.

**Comment by stephenh** [comment]
I merged #970 which addresses the file organization/structuring aspect of this issue.

@sheldonj per previous response, lmk if you have more specific asks/ideas on the circular references and `any` usage, but otherwise going to consider the "structure" aspect of this issue done for now.

Thanks!
